plasmapy/simulation/particle_tracker.py

```diff
@@ -0,0 +1,883 @@
+
+        """
+        if self.save_now or force_save:
+            self.save()
+
+
+class IntervalSaveRoutine(AbstractSaveRoutine):
+    """Abstract class describing a save routine that saves every given interval."""
+
+    def __init__(self, interval: u.Quantity, **kwargs):
+        super().__init__(**kwargs)
+        self.t_all = []
+
+        self.save_interval = interval.to(u.s).value
+        self.time_of_last_save = 0
+
+    @property
+    def require_synchronized_dt(self):
+        """Save output only makes sense for synchronized time steps."""
+        return True
+
+    @property
+    def save_now(self):
+        """Save at every interval given in instantiation."""
+
+        return self.tracker.time - self.time_of_last_save >= self.save_interval
+
+    def save(self):
+        """Save the current state of the simulation.
+        Sets the time of last save attribute and log the timestamp.
+        """
+        super().save()
+
+        self.time_of_last_save = self.tracker.time
+        self.t_all.append(self.tracker.time)
+
+
+class ParticleTracker:
+    r"""A many body particle tracker for electric and magnetic fields with no particle-particle interactions.
+
+    Particles are created and pushed through calculated E and B fields. These fields are specified as part of a grid
+    which can then be interpolated to determine local field strength. The time step used can be specified, otherwise an
+    adaptive time step is calculated based off grid resolution and the particle's gyroradius.
+
+    The simulation is terminated when a specified stop condition is met. The results of a simulation can be exported
+    using a save routine.
+
+    Parameters
+    ----------
+    grids : `~plasmapy.plasma.grids.AbstractGrid` or subclass thereof, or list
+        of same.
+        A Grid object or list of grid objects containing the required
+        quantities [E_x, E_y, E_z, B_x, B_y, B_z].
+        If any of these quantities are missing, a warning will be given and that
+        quantity will be assumed to be zero everywhere.
+
+    req_quantities : `list` of str, optional
+        A list of quantity keys required to be specified on the Grid object.
+        The base particle pushing simulation requires the quantities [E_x, E_y, E_z, B_x, B_y, B_z].
+        This keyword is for specifying quantities in addition to these six. The default is None.
+
+    verbose : bool, optional
+        If true, updates on the status of the program will be printed
+        into the standard output while running. The default is True.
+    """
+
+    def __init__(
+        self,
+        grids: Union[AbstractGrid, Iterable[AbstractGrid]],
+        req_quantities=None,
+        verbose=True,
+    ):
+        # self.grid is the grid object
+        if isinstance(grids, AbstractGrid):
+            self.grids = [
+                grids,
+            ]
+        elif isinstance(grids, collections.abc.Iterable):
```

Comment:
These elif and else statements aren't currently being hit in testing (according to codecov)